android/app/src/main/java/com/integreat/fetcher/DownloadResultCollector.java

```diff
@@ -0,0 +1,75 @@
+package com.integreat.fetcher;
+
+import android.support.annotation.Nullable;
+import com.facebook.react.bridge.*;
+import com.facebook.react.modules.core.DeviceEventManagerModule;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class DownloadResultCollector implements FileDownloadCallback {
+    private final Map<String, String> failedUrls = new HashMap<>();
+    private final Map<String, String> succeededUrls = new HashMap<>();
+    private final Promise promise;
+    private final List<String> expectedUrls;
+    private final ReactContext reactContext;
+
+    public DownloadResultCollector(ReactContext reactContext, List<String> expectedUrls, Promise promise) {
+        this.promise = promise;
+        this.expectedUrls = expectedUrls;
+        this.reactContext = reactContext;
+    }
+
+    @Override
+    public void failed(String url, IOException e) {
+        failedUrls.put(url, e.getMessage());
+        sendProgress();
+        tryToResolve();
+    }
+
+    @Override
+    public void downloaded(String url, File target) {
+        succeededUrls.put(url, target.getAbsolutePath());
+        sendProgress();
+        tryToResolve();
+    }
+
+    private void tryToResolve() {
+        if (failedUrls.size() + succeededUrls.size() != expectedUrls.size()) {
+            return;
+        }
+
+        WritableMap result = Arguments.createMap();
+
+        WritableMap failed = Arguments.createMap();
```

Comment:
Can't we just use a WritableMap from the start instead of a HashMap?

Reply:
WritableMap is a shitty class by react-native.This means it is very limited and does not implement the Map interface.

So its easier to work with a real Map.